openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+type IntrospectionPage struct {
+	pagination.LinkedPageBase
+}
+
+// Introspection represents an introspection in the OpenStack Bare Metal Introspector API.
+type Introspection struct {
+	// Whether introspection is finished
+	Finished bool `json:"finished"`
+
+	// State of the introspection
+	State string `json:"state"`
+
+	// Error message, can be null; "Canceled by operator" in case introspection was aborted
+	Error string `json:"error"`
+
+	// UUID of the introspection
+	UUID string `json:"uuid"`
+
+	// UTC ISO8601 timestamp
+	StartedAt time.Time `json:"-"`
+
+	// UTC ISO8601 timestamp or null
+	FinishedAt time.Time `json:"-"`
+
+	// Link to the introspection URL
+	Links []interface{} `json:"links"`
+}
+
+// IsEmpty returns true if a page contains no Server results.
+func (r IntrospectionPage) IsEmpty() (bool, error) {
+	s, err := ExtractIntrospections(r)
+	return len(s) == 0, err
+}
+
+// NextPageURL uses the response's embedded link reference to navigate to the
+// next page of results.
+func (r IntrospectionPage) NextPageURL() (string, error) {
+	var s struct {
+		Links []gophercloud.Link `json:"introspection_links"`
+	}
+	err := r.ExtractInto(&s)
+	if err != nil {
+		return "", err
+	}
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// UnmarshalJSON trie to convert values for started_at and finished_at from the
+// json response into RFC3339 standard. Since Introspection API can remove the
+// Z from the format, if the conversion fails, it falls back to an RFC3339
+// with no Z format supported by gophercloud.
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
+	}
+
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	*r = Introspection(s.tmp)
+
+	if s.StartedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.StartedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.StartedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.StartedAt = t
+	}
+
+	if s.FinishedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.FinishedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.FinishedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.FinishedAt = t
+	}
+
+	return nil
+}
+
+// GetIntrospectionStatusResult is the response from a Get operation.
+// Call its Extract method to interpret it as an Introspection.
+type GetIntrospectionStatusResult struct {
+	introspectionResult
+}
+
+// StartResult ...
+type StartResult struct {
+	gophercloud.ErrResult
+}
+
+// AbortResult ...
+type AbortResult struct {
+	gophercloud.ErrResult
+}
+
+// Data represents introspection data collected.
+// The format and contents of the stored data depends on the ramdisk used
+// and plugins enabled both in the ramdisk and in inspector itself.
+// This structure has been provided for basic compatibility but it
+// will need extensions
+type Data struct {
```

Comment:
I hope these fields are all optional, because we don't have any guarantees about it. Actually, maybe we should leave only `inventory` since it's the only thing that has a defined (in IPA) format.

Reply:
Please see my other comments in this area. I'm hesitant to have `Data` as one large `map[string]interface{}` for reasons mentioned. I'm a bit surprised if this API has no guaranteed / defined structure at all. 

However, as I've also stated, I have no knowledge of any of these APIs, so I'm leaning on other peoples' expertise. If it's best to have generic `map[string]interfaces{}` in certain areas, okay. But using the results is going to be a headache.

Also, if a JSON response does not contain a field that a struct is expecting, all is fine. If a JSON response contains _extra_ fields that the struct does not account for, all is fine.

Reply:
The problem with this data is that it's a hybrid of more or less well defined `inventory` and whatever ironic-inspector plugins put there (which is poorly documented and often legacy). Is it possible in Go to have a hybrid approach as well? I know in `serde` (Rust serializatin library) you can have a defined structure with one mapping field serving as catch-all for everything that was not mapped to defined fields.

Reply:
Go doesn't have anything which natively does that (but it would be nice). We have an internal function that does it, though. We use it for things like Keystone where arbitrary extra key/values can be sent in the API response:

https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/users/results.go#L60-L74

I've never tested it with nested things, but it _should_ work.

If we can get a confirmed set of base/expected fields, I'm more than happy to work on options to catch the undefined/optional pieces.